wsd/ClientSession.cpp

```diff
@@ -318,7 +318,59 @@ void ClientSession::handleClipboardRequest(DocumentBroker::ClipboardRequest
         if (data.get())
         {
             preProcessSetClipboardPayload(*data);
-            docBroker->forwardToChild(client_from_this(), "setclipboard\n" + *data, true);
+
+#if !MOBILEAPP
+            if (data->starts_with('{'))
+            {
+                // We got JSON, extract the URL and the UNO command name.
+                Poco::JSON::Object::Ptr json;
+                if (JsonUtil::parseJSON(*data, json))
+                {
+                    std::string url;
+                    JsonUtil::findJSONValue(json, "url", url);
+                    std::string commandName;
+                    JsonUtil::findJSONValue(json, "commandName", commandName);
+                    http::Session::FinishedCallback finishedCallback =
+                        [this, docBroker,
+                         commandName](const std::shared_ptr<http::Session>& session)
+                    {
+                        const std::shared_ptr<const http::Response> httpResponse =
+                            session->response();
+                        if (httpResponse->statusLine().statusCode() != http::StatusCode::OK)
+                        {
+                            LOG_ERR("Clipboard download request failed");
+                            return;
+                        }
+
+                        std::string body = httpResponse->getBody();
+                        docBroker->forwardToChild(client_from_this(), "setclipboard\n" + body,
```

Comment:
I'll make this more strict to require that the body is not a random format, but to some extent it's in the mimetype-length-data tuples format we expect here.